packages/gatsby/src/utils/page-data.ts

```diff
@@ -69,10 +79,83 @@ export async function writePageData(
   store.dispatch({
     type: `ADD_PAGE_DATA_STATS`,
     payload: {
-      filePath,
+      filePath: outputFilePath,
       size: pageDataSize,
     },
   })
 
-  await fs.outputFile(filePath, bodyStr)
+  await fs.outputFile(outputFilePath, bodyStr)
+  return body
+}
+
+let isFlushPending = false
+let isFlushing = false
+
+export function isFlushEnqueued(): boolean {
+  return isFlushPending
+}
+
+export async function flush(): Promise<void> {
+  if (isFlushing) {
+    // We're already in the middle of a flush
+    return
+  }
+  isFlushPending = false
+  isFlushing = true
+  const { pendingPageDataWrites, components, pages, program } = store.getState()
+
+  const { pagePaths, templatePaths } = pendingPageDataWrites
+
+  const pagesToWrite = Array.from(templatePaths).reduce(
+    (set, componentPath) => {
+      const templateComponent = components.get(componentPath)
+      if (templateComponent) {
+        templateComponent.pages.forEach(set.add.bind(set))
+      }
+      return set
+    },
+    new Set(pagePaths.values())
+  )
+
+  for (const pagePath of pagesToWrite) {
+    const page = pages.get(pagePath)
+
+    // It's a gloomy day in Bombay, let me tell you a short story...
+    // Once upon a time, writing page-data.json files were atomic
+    // After this change (#24808), they are not and this means that
+    // between adding a pending write for a page and actually flushing
+    // them, a page might not exist anymore щ（ﾟДﾟщ）
+    // This is why we need this check
+    if (page) {
+      const body = await writePageData(
+        path.join(program.directory, `public`),
+        page
+      )
+
+      if (program?._?.[0] === `develop`) {
+        websocketManager.emitPageData({
+          ...body.result,
+          id: pagePath,
+          result: {
+            data: body.result.data,
+            pageContext: body.result.pageContext,
+          },
+        })
+      }
+    }
+  }
+
+  store.dispatch({
+    type: `CLEAR_PENDING_PAGE_DATA_WRITES`,
+  })
+  isFlushing = false
+  return
+}
+
+export function enqueueFlush(): void {
```

Comment:
this name is misleading. it's not actually queuing. It's actually more of like a `flushIfReady` or something

Reply:
The thing is that it also sets the `isFlushPending` flag to `true` which the `isFlushEnqueued` fn is a getter for